client/driver/docker.go

```diff
@@ -355,35 +360,42 @@ func (d *DockerDriver) FSIsolation() cstructs.FSIsolation {
 	return cstructs.FSIsolationImage
 }
 
-func (d *DockerDriver) Prestart(ctx *ExecContext, task *structs.Task) error {
+func (d *DockerDriver) Prestart(ctx *ExecContext, task *structs.Task) (*CreatedResources, error) {
 	driverConfig, err := NewDockerDriverConfig(task, d.taskEnv)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
+	// Set state needed by Start()
+	d.driverConfig = driverConfig
+
 	// Initialize docker API clients
 	client, _, err := d.dockerClients()
 	if err != nil {
-		return fmt.Errorf("Failed to connect to docker daemon: %s", err)
+		return nil, fmt.Errorf("Failed to connect to docker daemon: %s", err)
 	}
 
-	if err := d.createImage(driverConfig, client, ctx.TaskDir); err != nil {
-		return err
+	// Ensure the image is available
+	pulled, err := d.createImage(driverConfig, client, ctx.TaskDir)
+	if err != nil {
+		return nil, err
 	}
+	if pulled {
+		// An image was downloaded, add the ID to created resources
+		dockerImage, err := client.InspectImage(driverConfig.ImageName)
+		if err == nil {
```

Comment:
This was intentional, but it is quite confusing to read.

Switched to the more common pattern.